src/bio.c

```diff
@@ -279,30 +290,76 @@ int wolfSSL_BIO_read(WOLFSSL_BIO* bio, void* buf, int len)
 }
 
 
+#ifdef WOLFSSL_BASE64_ENCODE
 /* Converts data into base64 output
  *
  * returns the resulting buffer size on success.
  */
+/**
+ * `out` buffer is allocated here and the caller is responsible
+ *       for free'ing it
+ */
 static int wolfSSL_BIO_BASE64_write(WOLFSSL_BIO* bio, const void* data,
-        word32 inLen, byte* out, word32* outLen)
+        word32 inLen, byte** out, word32* outLen, void* heap)
 {
+    int ret = WOLFSSL_FATAL_ERROR;
     byte* tmp = NULL;
-    int ret   = 0;
+    word32 sz = 0;
 
     WOLFSSL_ENTER("wolfSSL_BIO_BASE64_write");
 
     if (bio == NULL || data == NULL || out == NULL || outLen == NULL) {
         return BAD_FUNC_ARG;
     }
 
-#if defined(WOLFSSL_BASE64_ENCODE)
-    tmp = (byte*)XMALLOC(*outLen, bio->heap, DYNAMIC_TYPE_TMP_BUFFER);
-    if (tmp == NULL) {
+    /* get the encoded length */
+    if (bio->flags & WOLFSSL_BIO_FLAG_BASE64_NO_NL) {
+        if (Base64_Encode_NoNl((const byte*)data, inLen, NULL,
+                    &sz) != LENGTH_ONLY_E) {
+            WOLFSSL_MSG("Error with base64 get length");
+            return WOLFSSL_FATAL_ERROR;
+        }
+    }
+    else {
+        if (Base64_Encode((const byte*)data, inLen, NULL, &sz) !=
+            LENGTH_ONLY_E) {
+            WOLFSSL_MSG("Error with base64 get length");
+            return WOLFSSL_FATAL_ERROR;
+        }
+    }
+
+    if (sz == 0) {
+        return 0; /* nothing to do */
+    }
+
+    /* allocate buffer for encoded output */
+    if (*out == NULL) {
+        *out = (byte*)XMALLOC(sz, heap, DYNAMIC_TYPE_TMP_BUFFER);
+        if (*out == NULL) {
+            WOLFSSL_MSG("Memory error");
+            return WOLFSSL_FATAL_ERROR;
+    }
+        tmp = *out;
+    }
+    else {
+        if (sz > *outLen) {
+            /* use existing buffer as input */
+            *out = (byte*)XREALLOC(*out, sz, heap, DYNAMIC_TYPE_TMP_BUFFER);
+    if (*out == NULL) {
+        WOLFSSL_MSG("Memory error");
         return WOLFSSL_FATAL_ERROR;
     }
 
-    if ((bio->flags & WOLFSSL_BIO_FLAG_BASE64_NO_NL) ==
-            WOLFSSL_BIO_FLAG_BASE64_NO_NL) {
+        }
+        data = *out;
+        inLen = *outLen;
+        /* allocate temp buffer, since base64 encode does not allow inline */
+        tmp = (byte*)XMALLOC(sz, heap, DYNAMIC_TYPE_TMP_BUFFER);
```

Comment:
This buffer is always allocated so it can be moved before the `if` statement. In the `*out == NULL`, we would do `*out = tmp` and in the second case proceed as now but without the `tmp` buffer allocation.

Reply:
Not addressed.